File: backend/app/schemas/workflow.py
"""Pydantic contract for workflow run records."""

from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from app.domain.tasks import WorkflowEventType, WorkflowRunStatus


class WorkflowRunResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    run_id: UUID
    task_id: UUID
    thread_id: str
    graph_name: str
    graph_version: str
    status: WorkflowRunStatus
    started_at: datetime | None = None
    completed_at: datetime | None = None
    failed_at: datetime | None = None
    error_code: str | None = None
    error_message: str | None = Field(default=None, max_length=500)
    pending_action: str | None = None
    created_at: datetime
    updated_at: datetime


class WorkflowActionRequest(BaseModel):
    """统一 human action 入口（Stage 6A spec G）。

    - Stage 1 simulation run：approve_plan / cancel / retry；
    - Stage 5 真实研究 run：approve / rewrite / research / cancel；
    - comment 仅 Stage 5 human review 使用。
    """

    action_type: Literal["approve_plan", "cancel", "retry", "approve", "rewrite", "research"]
    comment: str | None = Field(default=None, max_length=500)


class WorkflowActionResponse(BaseModel):
    run: WorkflowRunResponse
    replayed: bool = False


class WorkflowEventResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    event_id: int
    run_id: UUID
    event_type: WorkflowEventType
    node_name: str | None = Field(default=None, max_length=64)
    stage: str | None = Field(default=None, max_length=64)
    progress: int | None = Field(default=None, ge=0, le=100)
    message: str = Field(max_length=500)
    payload: dict[str, object] = Field(default_factory=dict)
    created_at: datetime

File: frontend/src/types/workflow.ts
/** 与后端 /app/domain/tasks.py + /app/schemas/workflow.py 对齐。 */

export const WORKFLOW_RUN_STATUS = [
  'pending',
  'running',
  'waiting_human',
  'completed',
  'failed',
  'cancelled',
] as const;
export type WorkflowRunStatus = (typeof WORKFLOW_RUN_STATUS)[number];

export const WORKFLOW_EVENT_TYPE = [
  'run_created',
  'run_started',
  'node_completed',
  'run_completed',
  'run_failed',
  'run_waiting_human',
  'run_resumed',
  'run_cancelled',
] as const;
export type WorkflowEventType = (typeof WORKFLOW_EVENT_TYPE)[number];

/** Stage 1 simulation + Stage 5 真实研究的统一 human action。 */
export type ActionType =
  | 'approve_plan'
  | 'cancel'
  | 'retry'
  | 'approve'
  | 'rewrite'
  | 'research';

export interface WorkflowRunResponse {
  run_id: string;
  task_id: string;
  thread_id: string;
  graph_name: string;
  graph_version: string;
  status: WorkflowRunStatus;
  started_at: string | null;
  completed_at: string | null;
  failed_at: string | null;
  error_code: string | null;
  error_message: string | null;
  pending_action: string | null;
  created_at: string;
  updated_at: string;
}

export interface WorkflowActionRequest {
  action_type: ActionType;
  comment?: string | null;
}

export interface WorkflowActionResponse {
  run: WorkflowRunResponse;
  replayed: boolean;
}

export interface WorkflowEventResponse {
  event_id: number;
  run_id: string;
  event_type: WorkflowEventType;
  node_name: string | null;
  stage: string | null;
  progress: number | null;
  message: string;
  payload: Record<string, unknown>;
  created_at: string;
}
